### PR TITLE
Query optimization guide

### LINKED ISSUE #540: Document query optimizations
We need to document some query best practices for Flux. It should outline things like:

- Include the `filter()` and `group()` as early in the function as possible. This is related to "storage push-downs."
- Small window durations will always chew up resources.
- `map()`, `reduce()`, `join()`, `pivot()` and `union()` can be heavy and should only be used when absolutely necessary.

### PR DESCRIPTION
**Pull request by sanderson**
Closes #540

Just a guide for ways to optimize Flux queries.

- [x] Tests pass (no build errors)
- [x] Rebased/mergeable

### CODE REVIEW

File: content/v2.0/query-data/optimize-queries.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  v2_0:
+    name: Optimize queries
+    parent: Query data
+v2.0/tags: [query]
+---
+
+Optimize your Flux queries with a few simple principles.
+Query optimizations center around reducing the memory and compute (CPU) requirements
+of a query by changing function order, understanding how functions work, and being
+conscious of the nature of the data queried.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Optimize your Flux queries by reducing the memory and compute (CPU) that your queries require:
- Change the function order
- Understand how functions work (not sure this one explicitly reduces memory and compute?)
- Consider the nature of queried data (not sure what "nature" means here)

File: content/v2.0/query-data/optimize-queries.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+## Start queries with pushdown functions
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Scott, maybe replace the bullets above with headings:
- Start queries w...
- Increase window duration/intervals?
- Limit? "heavy" functions (maybe be explicit here: specified transformation functions?)

File: content/v2.0/query-data/optimize-queries.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+Certain Flux functions can push their data manipulation down to the underlying
+data source rather than pulling the data into memory and manipulating it there.
+Using "pushdown" functions reduces the amount of memory necessary to run a query.
+However, to benefit from these performance gains, you must **use pushdown functions
+at the beginning of your query**.
+When a non-pushdown function runs, Flux pulls the data into memory and manipulates the data there.
+All subsequent functions must operate in memory, including pushdown-capable functions.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Most? Flux functions manipulate data in memory. Flux pushdown functions reduce the memory required, but require data manipulation in an underlying data source? To increase query performance, you must **use pushdown functions
at the beginning of your query**.

File: content/v2.0/query-data/optimize-queries.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  |> range(start: -1h)                       //
+  |> filter(fn: (r) => r.sensor == "abc123") // Pushed to the data source
+  |> group(columns: ["_field", "host"])      //
+
+  |> aggregateWindow(every: 5m, fn: max)     //
+  |> filter(fn: (r) => r._value >= 90.0)     // Run in memory
+  |> top(n: 10)                              //
+```
+
+## Don't over-window data
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Increase window duration/intervals?

File: content/v2.0/query-data/optimize-queries.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+Windowing (grouping data based on time intervals) is commonly used to aggregate and downsample data.
+It's important to not "over-window" your data by using short window durations.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Improve performance by increasing...

File: content/v2.0/query-data/optimize-queries.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+The more windows Flux creates, the more compute power it needs to evaluate which
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: More windows require more compute to evaluate...

File: content/v2.0/query-data/optimize-queries.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+window each row should be assigned to.
+
+Reasonable window durations depend on the total time range queried.
+
+## Use "heavy" functions sparingly
+Some Flux functions are known to use more memory or CPU than others.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: The following transformation functions use more...consider how much these functions improve your data processing before using them:?

File: content/v2.0/query-data/optimize-queries.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+- [map()](/v2.0/reference/flux/stdlib/built-in/transformations/map/)
+- [reduce()](/v2.0/reference/flux/stdlib/built-in/transformations/aggregates/reduce/)
+- [window()](/v2.0/reference/flux/stdlib/built-in/transformations/window/)
+- [join()](/v2.0/reference/flux/stdlib/built-in/transformations/join/)
+- [union()](/v2.0/reference/flux/stdlib/built-in/transformations/union/)
+- [pivot()](/v2.0/reference/flux/stdlib/built-in/transformations/pivot/)
+
+{{% note %}}
+Flux engineers are in the process of optimizing functions.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: We're continually optimizing Flux.

File: content/v2.0/query-data/optimize-queries.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+This list may not represent the current state of Flux and will be updated over time.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: not sure if we want to say? may be inferred? "We'll continue to optimize Flux and update (the list of) Flux functions overtime."

File: content/v2.0/query-data/optimize-queries.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+{{% /note %}}
+
+## Balance time range vs data precision
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: vs >> and

File: content/v2.0/query-data/optimize-queries.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+To ensure queries are performant, be sure to balance the time range of your query
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: ...performant, balance...and the precision of your data

File: content/v2.0/query-data/optimize-queries.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+with the precision of your data.
+For example, if you query data with values stored every second and you request
+six months worth of data, results will include a minimum of ≈15.5 million points.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: is ~ the symbol here?

File: content/v2.0/query-data/optimize-queries.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+{{% note %}}
+Flux engineers are in the process of optimizing functions.
+This list may not represent the current state of Flux and will be updated over time.
+{{% /note %}}
+
+## Balance time range vs data precision
+To ensure queries are performant, be sure to balance the time range of your query
+with the precision of your data.
+For example, if you query data with values stored every second and you request
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: query data stored every second and request

File: content/v2.0/query-data/optimize-queries.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+six months worth of data, results will include a minimum of ≈15.5 million points.
+Flux has to store that data in memory as it generates a response.
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: Flux must store these points in memory to generate...

File: content/v2.0/query-data/optimize-queries.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+To query data over large periods of time, consider creating a task to
```
Comment (kelseiv) [inline content/v2.0/query-data/optimize-queries.md]: time, create a task to downsample data, and then query data...

File: content/v2.0/query-data/optimize-queries.md (hunk 12 above)
Comment (sanderson) [reply content/v2.0/query-data/optimize-queries.md]: The tilde `~` in mathematics means "proportional to." `≈` means "approximately."

**Pull request merged**